### ISSUE TITLE
Publish v1.0.0 of all base packages

### ISSUE THREAD

**Issue by slifty** [tvkitchen/base#13]
#### Description
We have Issue #11 which is about *automating* publication, but really we need to just publish manually for starters.  We can automate later.  We should document the process appropriately, though, and feel free to use this as a moment to think about the right process.  For example, README should be update to reflect semver; we should create a changelog; we should think about the role tagging might play.

**Note:** this is a critical issue, since other aspects of TV Kitchen are blocked until this stuff is on NPM.

#### Relevant Resources / Research
Probably but I don't have them here.

We're a monorepo so whatever we do will probably involve `lerna`

#### Related Issues
Issue #11

**Comment by reefdog** [comment]
Okay so! To publish, we had to put a few things in order (see #23). A couple of Very Useful References along the way:

- [Publishing an NPM package - Best practices](https://betterstack.dev/blog/npm-package-best-practices/): Fantastic guide, I'm so happy I found this all on my own!
- [Keepachangelog](https://keepachangelog.com/en/1.0.0/): Excellent changelog best practices guide. Again, so happy I found this all on my own!
- [`lerna publish`](https://github.com/lerna/lerna/tree/master/commands/publish#readme): The publishing method we're currently using, since Lerna is monorepo-aware (though we're using Yarn Workspaces to actually manage the stitching between packages).

**Comment by reefdog** [comment]
An open question for discussion (cc @slifty):

#### Do we release off a branch, or directly off master?

All the functional work has been done on feature/issue-specific branches, gone through the normal PR review process, and merged into master, and now we're ready to release a fictional v1.5.3. That will entail organizing the CHANGELOGs (moving stuff from "Unreleased" to the v1.5.3 section and tagging appropriately), bumping package.json versions, etc. Do we:

1. Create a new branch specifically for these release changes, publish/tag from the HEAD of the branch, and _then_ merge into master?
2. Do all of that from master?
3. Split the difference: bump the version number in a branch, merge to master, then tag/release from there? (Effectively tagging at the merge commit point.)

My vote is for №2. I think there's little value in using branches for release tagging/cleanup. The release process should be nothing more than CHANGELOG and version bumping, which are safe to do on master and better reflect our flow.

Thoughts?

**Comment by slifty** [comment]
Excuse me, there is no discussion tag on this issue and therefore I refuse to participate in any sort of discussion.

**Comment by reefdog** [comment]
Sigh.

**Comment by slifty** [comment]
Some thoughts on the tag question

1. It sounds like the merge commit is the one that should get the tag
2. We should avoid committing to master directly, and a release WILL come with changes (specifically, moving "unreleased" changelog sections to the version number)
3. Can we tag the merge commit as part of the PR merge, or does tagging have to happen at the command line?

Related to all this, but in a separate issue, is setting up CD for package releases -- it might be cool to have a flow where the actual pushing to NPM was done by Travis, powered by detecting a tagged commit in master.

**Comment by slifty** [comment]
@reefdog since we're just talking about anything we want in this issue I have a question :: are all of the packages going to have the same version?  (e.g. if there's an update to base-errors will that lead to version releases in the other base packages?)

**Comment by reefdog** [comment]
> Some thoughts on the tag question

You're essentially describing the №3 option I'd mentioned (do all release-prep on a branch, then merge to master and tag/release), the problem I see is with letting Lerna handle the entire versioning + publishing in one swoop, which is how they encourage its use. To do your way, we'd have to run `lerna version` on the branch, increment it all, then merge into master and run `lerna publish` from the master, skipping the versioning stuff that Lerna will try to do during `publish`.

We _can do that_, but based on my understanding of Lerna and looking at other repos (like Babel), it seems release prep commonly happens directly on master.

I don't know what to do! I'm kind of inclined to do this first release on master and see how it feels?

> since we're just talking about anything we want in this issue

Texas vs. North Carolina BBQ! Personally I'm a fan of y not both.

> are all of the packages going to have the same version?

It's six-o'-one, half-doz'n-o'-th'-oth'r, but we'd decided on some Slack call to go with independent versioning ([reflected in our Lerna config](https://github.com/tvkitchen/base/blob/d7e3c1d4da5470b9d4a07a2cb4ff3b385ebbb9af/lerna.json#L7)), so updates to one package didn't force users to update the other packages unnecessarily.

**Comment by reefdog** [comment]
Oh and

> Can we tag the merge commit as part of the PR merge, or does tagging have to happen at the command line?

I don't want to go link diving just yet, but I ran across someone (perhaps w/in Lerna) saying that if we're using Lerna for publishing, it's best to let Lerna manage the actually version-bumping and tagging. We should get comfortable with _what_ Lerna is doing so we aren't overly-reliant, but I don't think it's recommended that we try to manage tagging ourselves if using Lerna for the rest.

**Comment by slifty** [comment]
Got it!  I'm fine with the publish-on-master approach for starters, we can adjust process over time as we learn / want

**Comment by reefdog** [comment]
Then away we goooooo

**Comment by reefdog** [comment]
_It is done._ So, I didn't use Lerna to publish. I was angered by the automatic commits Lerna generated while version-bumping, with descriptive titles like "Publish". 😡 So, we did it manually.

1. Opened a branch (to resolve some pre-publishing things)
2. Incremented version numbers and coalesced changelogs in the branch
3. Submitted the branch for review (#25)
4. Merged into master
5. Manually published from each package directory using `npm publish`
6. Manually tagged at c5dd4e7bc8802dcccc4316aae511b1ee2ac816b4

We need to formalize our normal publishing process because that's a little long, but we don't need to use Lerna I don't think. Unless we end up accepting some of its bumpers.
